lectures/keras.md

```diff
@@ -0,0 +1,298 @@
+---
+jupytext:
+  text_representation:
+    extension: .md
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.16.4
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+# Simple Neural Network Regression with Keras and JAX
+
+```{include} _admonition/gpu.md
+```
+
+In this lecture we show how to implement one-dimensional nonlinear regression
+using a neural network.
+
+We will use the popular deep learning library [Keras](https://keras.io/), which
+provides a simple and elegant interface to deep learning.
+
+The emphasis in Keras on providing an intuitive API, while the heavy lifting is
+done by another library.
+
+Currently the backend library can be Tensorflow, PyTorch, or JAX.
+
+In this lecture we will use JAX.
+
+The objective of this lecture is to provide a very simple introduction to deep
+learning in a regression setting.
+
+We begin with some standard imports.
+
+```{code-cell} ipython3
+import numpy as np
+import matplotlib.pyplot as plt
+```
+
+Let's install Keras.
+
+```{code-cell} ipython3
+!pip install keras
+```
+
+Now we specify that the desired backend is JAX.
+
+```{code-cell} ipython3
+import os
+os.environ['KERAS_BACKEND'] = 'jax'
+```
+
+Next we import some tools from Keras.
+
+```{code-cell} ipython3
+import keras
+from keras.models import Sequential
+from keras.layers import Dense
+```
+
+```{code-cell} ipython3
+Dense?
+```
+
+## Data
+
+First let's write a function to generate some data.
+
+The data has the form
+
+$$
+    y_i = f(x_i) + \epsilon_i,
+    \qquad i=1, \ldots, n
+$$
+
+The map $f$ is specified inside the function and $\epsilon_i$ is an independent
+draw from a fixed normal distribution.
+
+Here's the function that creates vectors `x` and `y` according to the rule
+above.
+
+```{code-cell} ipython3
+def generate_data(x_min=0, x_max=5, data_size=400):
+    x = np.linspace(x_min, x_max, num=data_size)
+    x = x.reshape(data_size, 1)
+    ϵ = 0.2 * np.random.randn(*x.shape)
+    y = x**0.5 + np.sin(x) + ϵ
+    x, y = [z.astype('float32') for z in (x, y)]
+    return x, y
+```
+
+Now we generate some data to train the model.
+
+```{code-cell} ipython3
+x, y = generate_data()
+```
+
+Here's a plot of the training data.
+
+```{code-cell} ipython3
+fig, ax = plt.subplots()
+ax.scatter(x, y)
+ax.set_xlabel('x')
+ax.set_ylabel('y')
+plt.show()
+```
+
+We'll also use data from the same process for cross-validation.
+
+```{code-cell} ipython3
+x_validate, y_validate = generate_data()
+```
+
+## Models
+
+We supply functions to build two types of models.
+
+The first implements linear regression.
+
+This is achieved by constructing a neural network with just one layer, that maps
+to a single dimension (since the prediction
```

Comment:
@jstac this parentheses is incomplete.

Reply:
Thanks @mmcky , that was meant to be "(since the prediction is real-valued)"